node/src/components/linear_chain_sync/operations.rs

```diff
@@ -484,19 +522,170 @@ pub(crate) async fn run_fast_sync_task(
     }
 
     // Synchronize the trie store for the most recent block header.
-    info!(
-        state_root_hash = ?most_recent_block_header.state_root_hash(),
-        "Syncing trie store",
+    sync_trie_store(effect_builder, *most_recent_block_header.state_root_hash()).await?;
+
+    Ok((trusted_key_block_info, most_recent_block_header))
+}
+
+/// Downloads and saves the deploys and transfers for a block.
+async fn sync_deploys_and_transfers_and_state(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    block: &Block,
+) -> Result<(), LinearChainSyncError> {
+    for deploy_hash in block.deploy_hashes() {
+        fetch_and_store_deploy(effect_builder, *deploy_hash).await?;
+    }
+    for transfer_hash in block.transfer_hashes() {
+        fetch_and_store_deploy(effect_builder, *transfer_hash).await?;
+    }
+    sync_trie_store(effect_builder, *block.header().state_root_hash()).await?;
+    Ok(())
+}
+
+/// Archival sync all the way up to the current version.
+///
+/// Performs the following:
+///
+///  1. Fetches blocks all the way back to genesis.
+///  2. Fetches global state for each block starting from genesis onwards.
+///  3. Fetches all deploys and finality signatures.
+///  4. Stops after fetching a block with the current version and its global state.
+///
+/// Returns the block header with our current version and the last trusted key block information for
+/// validation.
+async fn archival_sync(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    trusted_block_header: BlockHeader,
+    chainspec: &Chainspec,
+) -> Result<(KeyBlockInfo, BlockHeader), LinearChainSyncError> {
+    // Get the trusted block info. This will fail if we are trying to join with a trusted hash in
+    // era 0.
+    let mut trusted_key_block_info =
+        get_trusted_key_block_info(effect_builder, chainspec, &trusted_block_header).await?;
+
+    let trusted_block =
+        *fetch_and_store_block_by_hash(effect_builder, trusted_block_header.hash()).await?;
+
+    // Sync to genesis
+    let mut walkback_block = trusted_block.clone();
+    loop {
+        sync_deploys_and_transfers_and_state(effect_builder, &walkback_block).await?;
+        if walkback_block.height() == 0 {
+            break;
+        } else {
+            walkback_block = *fetch_and_store_block_by_hash(
+                effect_builder,
+                *walkback_block.header().parent_hash(),
+            )
+            .await?;
+        }
+    }
+
+    // Sync forward until we are at the current version.
+    let mut most_recent_block = trusted_block;
+    while most_recent_block.header().protocol_version() < chainspec.protocol_config.version {
```

Comment:
We walk forward here, just like we do in `fast_sync_to_most_recent`, but the condition to stop looks different:
* there, we return an error on `if most_recent_block_header.protocol_version() > current_version`
* we finish when we couldn't download any more blocks AND we're in the current era AND we're on the same protocol version as the most recent block while here we keep downloading as long as our protocol version is higher than the most recent block's.

Other than that, the two pieces look very similar. We should extract it and reuse. Not only we get rid of code duplication but also we would have consistent logic. Unless I'm missing something.

Reply:
But the main part of the common functionality is already extracted in `fetch_and_store_block_by_height`; otherwise this is a rather small method and the differences you're listing (plus the call to `sync_deploys_and_transfers_and_state`, which `fast_sync_to_most_recent` doesn't do) covers pretty much all of it. I'm not sure whether trying to deduplicate this further would actually make the code simpler?

Reply:
At the very least, the fact that they both have different stopping conditions, is suspicious.

Reply:
That's a good point: Won't we fail to archive all those blocks that are at the current version but weren't fully downloaded by fast sync, @xcthulhu?

Reply:
Operating under the assumption that executing blocks is faster than downloading tries, this loop stops downloading tries after we reach a block with the current version.  After that, we download and execute blocks.